pkg/skaffold/docker/default_repo.go

```diff
@@ -17,41 +17,40 @@ limitations under the License.
 package docker
 
 import (
-	"regexp"
 	"strings"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/docker/registry"
 )
 
 const maxLength = 255
 
-var (
-	escapeRegex = regexp.MustCompile(`[/._:@]`)
-	prefixRegex = regexp.MustCompile(`(.*\.)?gcr.io/[a-zA-Z0-9-_]+/?`)
-)
-
-func SubstituteDefaultRepoIntoImage(defaultRepo string, image string) (string, error) {
+func SubstituteDefaultRepoIntoImage(defaultRepo string, image string, rewriteStrategy bool) (string, error) {
 	if defaultRepo == "" {
 		return image, nil
 	}
-
 	parsed, err := ParseReference(image)
 	if err != nil {
 		return "", err
 	}
 
-	replaced := replace(defaultRepo, parsed.BaseName)
+	var replaced string
+	if rewriteStrategy {
+		replaced = substituteDefaultRepoIntoImageWithRewrite(defaultRepo, parsed.BaseName)
+	} else {
+		replaced = replace(defaultRepo, parsed.BaseName)
+	}
```

Comment:
The truncation should happen here, before we add the tag and/or digest.
```suggestion
	}
	replaced = truncate(replaced)
```